src/compas/robots/base_artist/_artist.py

```diff
@@ -0,0 +1,309 @@
+    def draw_geometry(self, geometry, name=None, color=None):
+        """Draw a **COMPAS** geometry in the respective CAD environment.
+
+        Note
+        ----
+        This is an abstract method that needs to be implemented by derived classes.
+
+        Parameters
+        ----------
+        geometry : :class:`Mesh`
+            Instance of a **COMPAS** mesh
+        name : str, optional
+            The name of the mesh to draw.
+
+        Returns
+        -------
+        object
+            CAD-specific geometry
+        """
+        raise NotImplementedError
+
+
+class BaseRobotModelArtist(AbstractRobotModelArtist):
+    """Provides common functionality to most robot model artist implementations.
+
+    In **COMPAS**, the `artists` are classes that assist with the visualization of
+    datastructures and models, in a way that maintains the data separated from the
+    specific CAD interfaces, while providing a way to leverage native performance
+    of the CAD environment.
+
+    There are two methods that implementers of this base class should provide, one
+    is concerned with the actual creation of geometry in the native format of the
+    CAD environment (:meth:`draw_geometry`) and the other is one to apply a transformation
+    to geometry (:meth:`transform`).
+
+    Attributes
+    ----------
+    model : :class:`compas.robots.RobotModel`
+        Instance of a robot model.
+    """
+
+    def __init__(self, model):
+        super(BaseRobotModelArtist, self).__init__()
+        self.model = model
+        self.create()
+        self.scale_factor = 1.
+        self.attached_tool_model = None
+
+    def attach_tool_model(self, tool_model):
+        """Attach a tool to the robot artist.
+
+        Parameters
+        ----------
+        tool_model : :class:`compas.robots.ToolModel`
+            The tool that should be attached to the robot's flange.
+        """
+        self.attached_tool_model = tool_model
+        self.create(tool_model.root, 'attached_tool')
+
+        if not tool_model.link_name:
+            link = self.model.get_end_effector_link()
+            tool_model.link_name = link.name
+        else:
+            link = self.model.get_link_by_name(tool_model.link_name)
+
+        # originally, there was only this
+        ee_frame = link.parent_joint.origin.copy()
+        initial_transformation = Transformation.from_frame_to_frame(Frame.worldXY(), ee_frame)
+        # but if you attach a tool after applying some update to the robot, then it appears
+        # disconnected, being displayed at the original location of the end effector frame.
+        # I feel it shouldn't matter the order of calling attach_tool(_model) and update
+        # so I suggest the alternative, but maybe it isn't safe.
+
+        if link.collision:
+            relative_transformation = link.collision[0].current_transformation
+        else:
+            relative_transformation = link.visual[0].current_transformation
```

Comment:
As noted in the comment above, is this a terrible idea?

Reply:
So... this is a terrible idea? (The thumbs up response is very confusing.)

Reply:
no, I think it's fine. At least the explanation of the issue makes sense to me. Not sure if @romanarust has a different idea here...